flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcActor.java

```diff
@@ -183,6 +177,69 @@ protected void handleRpcMessage(Object message) {
 		}
 	}
 
+	private void handleControlMessage(ControlMessages controlMessage) {
+		switch (controlMessage) {
+			case START:
+				if (state == State.STOPPED) {
+					state = State.STARTED;
+				} else {
+					logStateTransitionWarning(state, State.STARTED);
+				}
+				break;
+			case STOP:
+				if (state == State.STARTED) {
+					state = State.STOPPED;
+				} else {
+					logStateTransitionWarning(state, State.STOPPED);
+				}
+				break;
+			case TERMINATE:
+				terminateRpcActor();
+				break;
+			default:
+				final String message = String.format("Received unknown control message %s. Dropping this message!", controlMessage);
+				log.warn(message);
+				sendErrorIfSender(new AkkaUnknownMessageException(message));
+		}
+	}
+
+	private void logStateTransitionWarning(State currentState, State targetState) {
```

Comment:
I think a more OO approach to this would be to define a state machine over the possible states that has the  set of valid transitions and will log validations if trying to perform an invalid transition. This would also give an explicit description of the valid transitions and the mental model behind it. The set of possible states here is rather small, but I observed a similar "problem" in the execution graph where the set of states and their transitions is more complex. I feel like this is a more maintainable and readable approach.

Reply:
I will wrap the logic in an inner class.